### PR TITLE
go-deadlock

### PR DESCRIPTION
**Pull request by henridf**
Introduce some instrumentation (https://github.com/sasha-s/go-deadlock/blob/master/deadlock.go)
that can be enabled to help detect potential deadlocks and diagnose actual deadlocks.

If the deadlock build tag is set, the deadlock sync.Mutex and
sync.RWMutex implementations are used. By default the tag is unset the
regular stdlib sync.Mutex and sync.RWMutex implementations are used as
previously.

### CODE REVIEW

**Comment by mattnibs** [comment]
I have yet to use this myself on anything else- the goroutine dump has always been enough for me. @nwt says neither has he. If I could truly not solve a deadlock issue I would probably just recreate this on a separate branch. As is, I'd prefer to import `sync` from the standard library.

**Comment by henridf** [comment]
It sounds like some of us have used it go-deadlock and and found it useful, while others haven't used it :)

> As is, I'd prefer to import sync from the standard library.

That's cool, but can you explain the reason for the preference? I can drop this if there's a reason to.

**Comment by mattnibs** [comment]
> I can drop this if there's a reason to.

This change adds an extra level of complexity and indirection, which is fine but the payout needs to justify the cost. The cost here is that reviewers need to make sure that all `sync` imports reference `pkg/sync` NOT stdlib `sync`. Even if we had a ci linting rule that checked for this, there would still be the cost of having to make changes code to fix the linting rule. (As an aside, will `goimport` always choose the correct `sync` package?)

As for the payout- for me the deadlocks where I conceivably would want to use are on deadlocks that are difficult to replicate; typically long running, load intensive type tests. I'd still have to configure those tests to run (potentially multiple times) on a binary built with the deadlock tag. At that point it wouldn't be much harder for me to create a branch and use `go-deadlock` on the portion of the code where deadlock is occurring (the goroutine stack dump will hint at the place to look).

tl;dr I do not believe the juice is worth the squeeze.

**Comment by henridf** [comment]
> extra level of complexity and indirection

For the record, the extra level of complexity and indirection is:

```go
type Mutex = sync.Mutex
type RWMutex = sync.RWMutex
```

> Even if we had a ci linting rule that checked for this

No CI linting is needed by this change.

**Comment by mattnibs** [comment]
> No CI linting is needed by this change.

I'm talking about how to enforce that new commits import from `pkg/sync` instead of `sync`.

**Comment by alfred-landrum** [comment]
We've used go-deadlock in other efforts to help us identify & diagnose deadlock conditions. At this early stage in zq, it seems like a net benefit; the extra step of using the correct package seems manageable, and using a build tag avoids any overhead except during testing. Even without a CI check to look for the standard 'sync', this is useful, as at the time we want to investigate deadlocks, we can make a manual pass looking for the sync package.

**Comment by henridf** [comment]
Closing as this has been controversial. We can revisit when we next run into hard deadlocks here or (more likely) upstream.

@nwt thanks for the tip; I didn't even know that gofmt could do this kind of thing.
